lib/ts/recipe/passwordless/api/implementation.ts

```diff
@@ -2,156 +2,810 @@ import { APIInterface } from "../";
+                    throw new SignInUpError({
                         status: "SIGN_IN_UP_NOT_ALLOWED",
                         reason:
                             "Cannot sign in / up due to security reasons. Please try a different login method or contact support. (ERR_CODE_003)",
-                    };
+                    });
                 }
+            };
 
-                // we do account linking only during sign in here cause during sign up,
-                // the recipe function above does account linking for us.
-                response.user = await AccountLinking.getInstance().createPrimaryUserIdOrLinkAccounts({
-                    tenantId: input.tenantId,
-                    user: response.user,
-                    userContext: input.userContext,
+            const linkAccountsForFactorSetup = async (
+                sessionUser: User,
+                recipeUserId: RecipeUserId,
+                userContext: UserContext
+            ) => {
+                if (!sessionUser.isPrimaryUser) {
+                    const createPrimaryRes = await AccountLinking.getInstance().recipeInterfaceImpl.createPrimaryUser({
+                        recipeUserId: new RecipeUserId(sessionUser.id),
+                        userContext,
+                    });
+                    if (createPrimaryRes.status === "RECIPE_USER_ID_ALREADY_LINKED_WITH_PRIMARY_USER_ID_ERROR") {
+                        // Session user is linked to another primary user, which means the session is revoked as well
+
+                        // We cannot recurse here because when the session user if fetched again,
+                        // it will be a primary user and we will end up trying factor setup with that user
+                        // Also this session would have been revoked and we won't be able to catch it again
+
+                        throw new SessionError({
+                            type: SessionError.UNAUTHORISED,
+                            message: "Session may be revoked",
+                        });
+                    } else if (
+                        createPrimaryRes.status === "ACCOUNT_INFO_ALREADY_ASSOCIATED_WITH_ANOTHER_PRIMARY_USER_ID_ERROR"
+                    ) {
+                        throw new RecurseError();
+                    }
+                }
+
+                const linkRes = await AccountLinking.getInstance().recipeInterfaceImpl.linkAccounts({
+                    recipeUserId: recipeUserId,
+                    primaryUserId: sessionUser.id,
+                    userContext,
                 });
-            }
 
-            const session = await Session.createNewSession(
-                input.options.req,
-                input.options.res,
-                input.tenantId,
-                loginMethod.recipeUserId,
-                {},
-                {},
-                input.userContext
-            );
+                if (linkRes.status !== "OK") {
+                    throw new RecurseError();
```

Comment:
add a comment here similar to emailpassword.

Reply:
Done